From b7932630923b6e4f0e16b7a0991ff573a4cab4cb Mon Sep 17 00:00:00 2001
From: tpatzig <tpatzig@283d02a7-25f6-0310-bc7c-ecb5cbfe19da>
Date: Fri, 16 Jan 2009 13:56:00 +0000
Subject: [PATCH] - add adapter to deviceMap on adapter_added Signal
 - emit signal defaultAdapterChanged
 - take removed adapters from the backend Map

git-svn-id: svn://anonsvn.kde.org/home/kde/trunk/KDE/kdebase@912004 283d02a7-25f6-0310-bc7c-ecb5cbfe19da
---
 .../libs/solid/control/bluetoothinterface.cpp      |    1 -
 workspace/libs/solid/control/bluetoothmanager.cpp  |   25 ++++++++++++++++++-
 workspace/libs/solid/control/bluetoothmanager.h    |    1 +
 .../libs/solid/control/ifaces/bluetoothmanager.h   |    7 +++++
 4 files changed, 31 insertions(+), 3 deletions(-)

diff --git a/workspace/libs/solid/control/bluetoothinterface.cpp b/workspace/libs/solid/control/bluetoothinterface.cpp
index fc562b2..e61d1d0 100644
--- a/workspace/libs/solid/control/bluetoothinterface.cpp
+++ b/workspace/libs/solid/control/bluetoothinterface.cpp
@@ -536,7 +536,6 @@ void Solid::Control::BluetoothInterfacePrivate::setBackendObject(QObject *object
 */
         QObject::connect(object, SIGNAL(deviceCreated(const QString &)),
                                  parent(), SIGNAL(deviceCreated(const QString &)));
-
         QObject::connect(object, SIGNAL(deviceDisappeared(const QString &)),
                                  parent(), SIGNAL(deviceDisappeared(const QString &)));
         QObject::connect(object, SIGNAL(deviceFound(const QString &, const QMap<QString,QVariant> &)),
diff --git a/workspace/libs/solid/control/bluetoothmanager.cpp b/workspace/libs/solid/control/bluetoothmanager.cpp
index 7a4c114..551ae9d 100644
--- a/workspace/libs/solid/control/bluetoothmanager.cpp
+++ b/workspace/libs/solid/control/bluetoothmanager.cpp
@@ -58,6 +58,7 @@ public:
 
     void _k_interfaceAdded(const QString &ubi);
     void _k_interfaceRemoved(const QString &ubi);
+    void _k_defaultInterfaceChanged(const QString &ubi);
     void _k_interfaceDestroyed(QObject *object);
 /*
     void _k_inputDeviceCreated(const QString &ubi);
@@ -110,6 +111,8 @@ Solid::Control::BluetoothInterfaceList Solid::Control::BluetoothManager::buildDe
 
     if (backend == 0) return list;
 
+    kDebug() << "UBI List " << ubiList;
+
     foreach (const QString &ubi, ubiList) {
         QPair<BluetoothInterface *, Ifaces::BluetoothInterface *> pair = d->findRegisteredBluetoothInterface(ubi);
 
@@ -239,7 +242,9 @@ void Solid::Control::BluetoothManager::removeInputDevice(const QString &ubi)
 */
 void Solid::Control::BluetoothManagerPrivate::_k_interfaceAdded(const QString &ubi)
 {
-    QPair<BluetoothInterface *, Ifaces::BluetoothInterface *> pair = bluetoothInterfaceMap.take(ubi);
+    kDebug() << "Size of InterfaceList " << bluetoothInterfaceMap.size();
+    QPair<BluetoothInterface *, Ifaces::BluetoothInterface *> pair = findRegisteredBluetoothInterface(ubi);
+/*    QPair<BluetoothInterface *, Ifaces::BluetoothInterface *> pair = bluetoothInterfaceMap.take(ubi);
 
     if (pair.first != 0) {
         // Oops, I'm not sure it should happen...
@@ -247,7 +252,7 @@ void Solid::Control::BluetoothManagerPrivate::_k_interfaceAdded(const QString &u
 
         delete pair.first;
         delete pair.second;
-    }
+    }*/
 
     emit q->interfaceAdded(ubi);
 }
@@ -261,11 +266,19 @@ void Solid::Control::BluetoothManagerPrivate::_k_interfaceRemoved(const QString
         delete pair.second;
     }
 
+    Ifaces::BluetoothManager *backend = qobject_cast<Ifaces::BluetoothManager *>(managerBackend());
+    backend->removeInterface(ubi);
     emit q->interfaceRemoved(ubi);
 }
 
+void Solid::Control::BluetoothManagerPrivate::_k_defaultInterfaceChanged(const QString &ubi)
+{
+    emit q->defaultInterfaceChanged(ubi);
+}
+
 void Solid::Control::BluetoothManagerPrivate::_k_interfaceDestroyed(QObject *object)
 {
+    kDebug() << "Interface detroyed";
     Ifaces::BluetoothInterface *device = qobject_cast<Ifaces::BluetoothInterface *>(object);
 
     if (device != 0) {
@@ -323,6 +336,9 @@ void Solid::Control::BluetoothManagerPrivate::connectBackend(QObject *newBackend
                      q, SLOT(_k_interfaceAdded(const QString &)));
     QObject::connect(newBackend, SIGNAL(interfaceRemoved(const QString &)),
                      q, SLOT(_k_interfaceRemoved(const QString &)));
+    QObject::connect(newBackend, SIGNAL(defaultInterfaceChanged(const QString &)),
+                     q, SLOT(_k_defaultInterfaceChanged(const QString &)));
+
 /*
     QObject::connect(newBackend, SIGNAL(inputDeviceCreated(const QString &)),
                      q, SLOT(_k_inputDeviceCreated(const QString &)));
@@ -334,17 +350,22 @@ void Solid::Control::BluetoothManagerPrivate::connectBackend(QObject *newBackend
 
 QPair<Solid::Control::BluetoothInterface *, Solid::Control::Ifaces::BluetoothInterface *> Solid::Control::BluetoothManagerPrivate::findRegisteredBluetoothInterface(const QString &ubi) const
 {
+
+    kDebug() << "findRegisteredBluetoothInterface " << ubi;
     if (bluetoothInterfaceMap.contains(ubi)) {
         return bluetoothInterfaceMap[ubi];
     } else {
+        kDebug() << "Creating New Interface " << ubi;
         Ifaces::BluetoothManager *backend = qobject_cast<Ifaces::BluetoothManager *>(managerBackend());
         Ifaces::BluetoothInterface *iface = 0;
 
         if (backend != 0) {
+            kDebug() << "Calling Backend to Creating New Interface " << ubi;
             iface = qobject_cast<Ifaces::BluetoothInterface *>(backend->createInterface(ubi));
         }
 
         if (iface != 0) {
+            kDebug() << "BackendIface created ";
             BluetoothInterface *device = new BluetoothInterface(iface);
             QPair<BluetoothInterface *, Ifaces::BluetoothInterface *> pair(device, iface);
             QObject::connect(iface, SIGNAL(destroyed(QObject *)),
diff --git a/workspace/libs/solid/control/bluetoothmanager.h b/workspace/libs/solid/control/bluetoothmanager.h
index 6864677..d2999af 100644
--- a/workspace/libs/solid/control/bluetoothmanager.h
+++ b/workspace/libs/solid/control/bluetoothmanager.h
@@ -188,6 +188,7 @@ private:
 
     Q_PRIVATE_SLOT(d, void _k_interfaceAdded(const QString &))
     Q_PRIVATE_SLOT(d, void _k_interfaceRemoved(const QString &))
+    Q_PRIVATE_SLOT(d, void _k_defaultInterfaceChanged(const QString &))
     Q_PRIVATE_SLOT(d, void _k_interfaceDestroyed(QObject *))
 /*
     Q_PRIVATE_SLOT(d, void _k_inputDeviceCreated(const QString &))
diff --git a/workspace/libs/solid/control/ifaces/bluetoothmanager.h b/workspace/libs/solid/control/ifaces/bluetoothmanager.h
index 6df0790..207e61e 100644
--- a/workspace/libs/solid/control/ifaces/bluetoothmanager.h
+++ b/workspace/libs/solid/control/ifaces/bluetoothmanager.h
@@ -89,6 +89,13 @@ public:
      virtual QObject *createInterface(const QString &ubi) = 0;
 
     /**
+     * Removes a BluetoothInterface object from this backend given its UBI.
+     *
+     * @param ubi the identifier of the bluetooth interface instantiated
+     */
+     virtual void removeInterface(const QString &ubi) = 0;
+
+    /**
      * Retrieves the list of Universal Bluetooth Identifiers (UBIs) of bluetooth input devices
      * which are configured in the system. Configured means also not connected devices.
      *
-- 
1.6.1

